File: Residential_Controller.py
import time
WaitingTime = 1


class ElevatorController:
    def __init__(self, NumberOfFloors, NumberOfElevators):
        self.NumberOfFloors = NumberOfFloors
        self.NumberOfElevators = NumberOfElevators
        self.column = Column(NumberOfFloors, NumberOfElevators)
        print("Controller iniatiated")

#------------------------Request Elevator---------------------------------------------------------------
    def RequestElevator(self, currentFloor, currentDirection,WaitingTime):
        time.sleep(WaitingTime)
        print("---------------------------------------------------")
        print("Request elevator to floor : ", currentFloor)
        time.sleep(WaitingTime)
        print("---------------------------------------------------")
        print("Call Button Light On")
        time.sleep(WaitingTime)
        elevator = self.find_best_elevator(currentFloor, currentDirection)
        elevator.send_request(currentFloor,WaitingTime)
        return elevator

#-----------------------Request Floor----------------------------------------------------------------
    def RequestFloor(self, elevator, RequestFloor,WaitingTime):
        time.sleep(WaitingTime)
        print("---------------------------------------------------")
        print("Requested floor : ", RequestFloor)
        time.sleep(WaitingTime)
        print("---------------------------------------------------")
        print("Request Button Light On")
        time.sleep(WaitingTime)
        elevator.send_request(RequestFloor,WaitingTime)

#  -------------Find the Best elevator-----------------------------------------------------

    def find_best_elevator(self, currentFloor, currentDirection):
        bestElevator = None
        BestDistance = 1000
        for elevator in (self.column.elevatorList):
            if (currentFloor == elevator.elevator_floor and (elevator.status == "stopped" or elevator.status == "idle" or elevator.status == "moving")):
                return elevator
            else:
                newDistance = abs(currentFloor - elevator.elevator_floor)
                if BestDistance > newDistance:
                    BestDistance = newDistance
                    bestElevator = elevator

                elif elevator.currentDirection == currentDirection:
                    bestElevator = elevator

        return bestElevator

##---------------------------------------------------------------------------------------

class Elevator:
    def __init__(self, ElevatorN, status, elevator_floor, elevator_currentDirection):
        self.ElevatorN = ElevatorN
        self.status = status
        self.elevator_floor = elevator_floor
        self.elevator_currentDirection = elevator_currentDirection
        self.floor_list = []

    def send_request(self, RequestFloor,waitingTime):
        self.floor_list.append(RequestFloor)
        self.addFloorList()
        self.moveElevator(RequestFloor,waitingTime)


    def addFloorList(self):
        if self.elevator_currentDirection == "up":
            self.floor_list.sort()
        elif self.elevator_currentDirection == "down":
            self.floor_list.sort()
            self.floor_list.reverse()
        return self.floor_list


    def moveElevator(self, RequestFloor,waitingTime):
        while (len(self.floor_list) > 0):
            if ((RequestFloor == self.elevator_floor)):
                self.openDoors(waitingTime)
                self.status = "moving"

                self.floor_list.pop()
            elif (RequestFloor < self.elevator_floor):

                self.status = "moving"
                print("---------------------------------------------------")
                print("Elevator", self.ElevatorN, self.status)
                print("---------------------------------------------------")
                self.currentDirection = "down"
                self.Move_down(RequestFloor,1)
                self.status = "stopped"
                print("---------------------------------------------------")
                print("Elevator", self.ElevatorN, self.status)
                print("---------------------------------------------------")
                self.openDoors(RequestFloor)
                self.floor_list.pop()
            elif (RequestFloor > self.elevator_floor):

                time.sleep(WaitingTime)
                self.status = "moving"
                print("---------------------------------------------------")
                print("Elevator", self.ElevatorN, self.status)
                print("---------------------------------------------------")
                self.currentDirection = "up"
                self.Move_up(RequestFloor,waitingTime)
                self.status = "stopped"
                print("---------------------------------------------------")
                print("Elevator", self.ElevatorN, self.status)
                print("---------------------------------------------------")

                self.openDoors(WaitingTime)

                self.floor_list.pop()

        if self.floor_list == 0:
            self.status = "idle"


# ------------- open Doors// close Doors -----------------------------------------------------------------------------

    def openDoors(self,waitingTime):
        time.sleep(WaitingTime)
        print("Open Door")
        print("---------------------------------------------------")
        print("Button Light Off")
        time.sleep(WaitingTime)
        print("---------------------------------------------------")
        time.sleep(WaitingTime)
        self.closeDoors(waitingTime)

    def closeDoors(self,waitingTime):
        print("close door")
        time.sleep(WaitingTime)


# -------------MOVE  ELEVATOR---------------------------------------------------------------------------------

    def Move_up(self, RequestFloor,WaitingTime):
        print("Floor : ", self.elevator_floor)
        time.sleep(WaitingTime)
        while(self.elevator_floor != RequestFloor):
            self.elevator_floor += 1
            print("Floor : ", self.elevator_floor)
            time.sleep(WaitingTime)

    def Move_down(self, RequestFloor,WaitingTime):
        print("Floor : ", self.elevator_floor)
        time.sleep(WaitingTime)
        while(self.elevator_floor != RequestFloor):
            self.elevator_floor -= 1
            print("Floor : ", self.elevator_floor)
            time.sleep(WaitingTime)


class Call_button:
    def __init__(self, currentFloor, currentDirection):
        self.currentFloor = currentFloor
        self.currentDirection = currentDirection
        self.light = False


class Floor_button:
    def __init__(self, RequestFloor):
        self.RequestFloor = RequestFloor


class Column:
    def __init__(self, NumberOfFloors, NumberOfElevators):
        self.NumberOfFloors = NumberOfFloors
        self.NumberOfElevators = NumberOfElevators
        self.elevatorList = []
        for i in range(NumberOfElevators):
            elevator = Elevator(i, "idle", 1, "up")
            self.elevatorList.append(elevator)


# // // //------------- TEST - -------------// // //

controller = ElevatorController(10, 2)
controller.column.elevatorList[0].elevator_floor = 2

controller.column.elevatorList[0].status = "moving"
controller.column.elevatorList[0].elevator_currentDirection = "down"
controller.column.elevatorList[1].elevator_floor = 6

controller.column.elevatorList[1].status = "moving"
Controller.column.elevatorList[1].elevator_currentDirection = "down"

elevator = controller.RequestElevator(5, "up",1)
controller.RequestFloor(elevator, 7,1)
print("==============================")
print(" End Senario 1 ")
print("==============================")


# // // //------------- WORKING - -------------// // //
controller.column.elevatorList[0].elevator_floor = 10
controller.column.elevatorList[0].status = "moving"
controller.column.elevatorList[0].elevator_currentDirection = "down"
controller.column.elevatorList[1].elevator_floor = 3
controller.column.elevatorList[1].status = "moving"controller.column.elevatorList[1].elevator_currentDirection = "down"

elevator = controller.RequestElevator(1, "up",1)
controller.RequestFloor(elevator, 6,1)
elevator = controller.RequestElevator(3, "up",1)
controller.RequestFloor(elevator, 5,1)
elevator = controller.RequestElevator(9, "down",1)
controller.RequestFloor(elevator, 2,1)
print("==============================")
print("End Senario 2")
print("==============================")

#// // //------------- WORKINGGGG - -------------//// //

controller = ElevatorController(10, 2)

controller.column.elevatorList[0].elevator_floor = 10
controller.column.elevatorList[0].status = "moving"
controller.column.elevatorList[0].elevator_currentDirection = "down"
controller.column.elevatorList[1].elevator_floor = 3
controller.column.elevatorList[1].status = "moving"
controller.column.elevatorList[1].elevator_currentDirection = "down"

print(controller.column.elevatorList)
elevator = controller.RequestElevator(10, "down",1)
controller.RequestFloor(elevator, 3,1)

elevator = controller.RequestElevator(3, "down",1)
controller.RequestFloor(elevator, 2,1)
print("==============================")
print("End Senario 3")
print("==============================")

File: Residential_Controller.js


function ElevatorController(NumberOfFloors, NumberOfElevators) {
    var syst = new ElevatorController(NumberOfFloors, NumberOfElevators);

    return syst;
}

class Column {
    constructor(NumberOfFloors, NumberOfElevators) {

        this.NumberOfFloors = NumberOfFloors;
        this.NumberOfElevators = NumberOfElevators;
        this.elevatorList = [];

        for (let i = 0; i < this.NumberOfElevators; i++) {

            let elevator = new Elevator(i + 1, "Idle", 1, "Up");
            this.elevatorList.push(elevator);
        }
    }
}

class Elevator {

    constructor(elevatorN, status, elevator_currentfloor, elevator_currentDirection) {

        this.elevatorN = elevatorN;
        this.status = status;
        this.elevator_floor = elevator_floor;
        this.elevator_currentDirection = elevator_currentDirection;
        this.floor_list = [];
    }

    addFloorList(RequestFloor) {

        this.floor_list.push(RequestFloor);
        this.sortFloorList();
        this.moveElevator(RequestFloor);
    }

    sortFloorList() {
        if (this.elevator_currentDirection === "Up") {
            this.floor_list.sort();
        } else if (this.elevator_currentDirection === "Down") {
            this.floor_list.sort();
            this.floor_list.reverse();
        }
        return this.floor_list;
    }

//-------------MOVE  ELEVATOR---------------------------------------------------------------------------------

    moveElevator(RequestFloor) {
        while (this.floor_list > 0) {
            if (RequestFloor === this.elevator_floor) {
                this.openDoors();
                this.status = "Moving";

                this.floor_list.shift();
            } else if (RequestFloor < this.elevator_floor) {
                this.status = "Moving";
                console.log("Elevator " + this.elevatorN + " is selected");
                console.log("Elevator " + this.elevatorN, this.status);
                console.log("");
                this.Direction = "Down";
                this.moveDown(RequestFloor);
                this.status = "Stopped";
                console.log("");
                console.log("Elevator " + this.elevatorN, this.status);
                console.log("");
                this.openDoors();
                this.floor_list.shift();
            } else if (RequestFloor > this.elevator_floor) {
                time(1000);
                this.status = "Moving";
                console.log("");
                console.log("Elevator " + this.elevatorN + " is selected");
                console.log("Elevator " + this.elevatorN, this.status);
                console.log("");
                this.Direction = "Up";
                this.moveUp(RequestFloor);
                this.status = "Stopped";
                console.log("");
                console.log("Elevator " + this.elevatorN, this.status);
                console.log("");

                this.openDoors();

                this.floor_list.shift();
            }
        }
        if (this.floor_list === 0) {
            this.status = "Idle";
        }
    }


    Request_floor_button(RequestFloor) {

        this.RequestFloor = RequestFloor;
        this.floorLights = floorLights;
    }

    Call_floor_button(currentfloor,  currentDirection) {

        this.currentfloor = currentfloor;
        this.currentDirection = currentDirection;
    }

// ------------- open Doors// close Doors -----------------------------------------------------------------------------

    openDoors() {
        time(1000);
        console.log("Doors opening");
        console.log("Doors opened");
        console.log("");
        console.log("Button light-off");
        time(1000);

        console.log("");
        time(1000);
        this.Close_door();
    }
    Close_door() {
        console.log("Closing doors");
        console.log("Doors' not obstructed : OK")
        console.log("Elevator's capacity check: OK ")
        console.log("Doors Closed")
        time(1000);
    }

// -------------MOVE  ELEVATOR---------------------------------------------------------------------------------


    moveUp(RequestFloor) {
        console.log("Floor : " + this.elevator_floor);
        time(1000);
        while (this.elevator_floor !== RequestFloor) {
            this.elevator_floor += 1;
            console.log("Floor : " + this.elevator_floor);

            time(1000);
        }
    }

    moveDown(RequestFloor) {
        console.log("Floor : " + this.elevator_floor);
        time(1000);
        while (this.elevator_floor !== RequestFloor) {
            this.elevator_floor -= 1;
            console.log("Floor : " + this.elevator_floor);

            time(1000);
        }
    }
}


class ElevatorController {
    constructor(NumberOfFloors, NumberOfElevators) {
        this.NumberOfFloors = NumberOfFloors;
        this.NumberOfElevators = NumberOfElevators;
        this.column = new Column(NumberOfFloors, NumberOfElevators);


        console.log("System is starting");
    }


//------------------------Request Elevator---------------------------------------------------------------

    RequestElevator(currentfloor, currentDirection) {
        time(1000);
        console.log("");
        console.log("Request elevator to floor : ", currentfloor);
        time(1000);
        console.log("");
        console.log("Call button illuminated");
        time(1000);

        let elevator = this.Find_Best_elevetor(currentfloor, currentDirection);
        elevator.addFloorList(floor);
        return elevator;
    }

// ----------------------Request Floor-------------------------------------------------------------------

    RequestFloor(elevator, RequestFloor) {
        time(1000);
        console.log("");
        console.log("The floor number " + RequestFloor + " is requested");
        time(1000);
        console.log("");
        console.log("Request button illuminated");
        time(1000);
        elevator.addFloorList(RequestFloor);

    }
    
    
//  -------------Find the Best elevator-----------------------------------------------------
    
    Find_Best_elevetor(currentfloor, currentDirection) {

        let Bestelevator = null;
        let Bestdistance = 1000;
        for (let i = 0; i < this.column.elevatorList.length; i++) {
            let elevator = this.column.elevatorList[i];

            if (floor === elevator.elevator_floor && (elevator.status === "Stopped" || elevator.status === "Idle" || elevator.status === "Moving")) {
                return elevator;
            } else {
                let newdistance = Math.abs(floor - elevator.elevator_floor);
                if (shortestDistance > newdistance) {
                    Bestdistance = newdistance;
                    Bestelevator = elevator;

                    if (elevator.Direction === Direction) {
                        Bestelevator = elevator;
                    }
                }
            }
        }
        returnBestelevator;
    }
}
 //-------------WaitingTime-----------------------------------------------------

function time( WaitingTime ) {
    var start = new Date().getTime();
    for (var i = 0; i < 1e7; i++) {
        if (new Date().getTime() - start >  WaitingTime ) {
            break;
        }
    }
}




// -----------------Test-----------------------------------------------------

function TestElevator() {

    var Scenario = initSystem(10, 2);
    var elevatorOne = Scenario.column.elevatorList[0];
    var elevatorTwo = Scenario.column.elevatorList[1];


    //Scenario 1 

    elevatorOne.currentfloor = 2;                           
    elevatorOne.status = "Idle";
    elevatorOne.currentDirection = "Up";

    elevatorTwo.currentfloor = 6;                           
    elevatorTwo.status = "Idle";
    elevatorTwo.currentDirection = "Down";

    var elevator = Scenario.RequestElevator(3, "Up");       
    Scenario.RequestFloor(elevator, 7);                       

    console.log("")

    console.log("")
    console.log("---Test scenario 1 : Ok++")                
    console.log("")
    console.log("+++")

    //Scenario 2 

    elevatorOne.currentfloor = 10;     
    elevatorOne.status = "Idle";
    elevatorOne.currentDirection = "Up";

    elevatorTwo.currentfloor = 3;                    
    elevatorTwo.status = "Idle";
    elevatorTwo.currentDirection = "Up";

    var elevator = Scenario.RequestElevator(1, "Up");        
    Scenario.RequestFloor(elevator, 6);                      

    elevator = Scenario.RequestElevator(3, "Up");            
    Scenario.RequestFloor(elevator, 5);                     

    elevator = Scenario.RequestElevator(9, "Down");            
    Scenario.RequestFloor(elevator, 2);                       

    console.log("");
    console.log("---Test scenario 2 : Ok++");
    console.log("");
    console.log("+++")

//Scenario 3

    elevatorOne.currentfloor = 10;
    elevatorOne.status = "Idle";
    elevatorOne.currentdirection = "Down";

    elevatorTwo.currentfloor = 3;
    elevatorTwo.status = "Moving";
    elevatorTwo.currentdirection = "Up";


    var elevator = Scenario.RequestElevator(3, "Down");

    Scenario.RequestFloor(elevator, 2);
    elevator = Scenario.RequestElevator(10, "Down");
    Scenario.RequestFloor(elevator, 3);


    console.log("");
    console.log("---Test scenario 3 : Ok++")
    console.log("");
    console.log("+++")
